### PR TITLE
QUIC transport + API changes

### LINKED ISSUE #44: Feature: QUIC transport
## Context
Implement QUIC transport using one of the following crates:
- https://github.com/aws/s2n-quic
- https://github.com/quinn-rs/quinn

Will need to think about multiplexing here, ideally abstracted away.

### PR DESCRIPTION
**Pull request by mempirate**
refactor(transport): extract TCP transport to module

refactor: rework socket interface with generic transport

wip: quic

Closes #44

### CODE REVIEW

**Comment by mempirate** [comment]
Current dependencies on/for this PR:
* `main`
  * **PR #48** <a href="https://app.graphite.dev/github/pr/chainbound/msg-rs/48?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/chainbound/msg-rs/48?utm_source=stack-comment).

File: msg-socket/src/pub/socket.rs [hunk 1]
```diff
@@ -61,22 +66,24 @@ impl<T: ServerTransport> PubSocket<T> {
     }
 
     /// Binds the socket to the given address. This spawns the socket driver task.
-    pub async fn bind(&mut self, addr: &str) -> Result<(), PubError> {
-        // Take the transport here, so we can move it into the backend task
-        let mut transport = self.transport.take().unwrap();
-
+    pub async fn bind(&mut self, addr: SocketAddr) -> Result<(), PubError> {
```
Comment (merklefruit) [inline msg-socket/src/pub/socket.rs:69]: nit: what do you think of following the tokio convention of using the generic `ToSocketAddrs` trait?
It's the one used in TcpStream::connect:

```rust
pub async fn connect<A: ToSocketAddrs>(addr: A) -> io::Result<TcpStream> {
    let addrs = to_socket_addrs(addr).await?;

    let mut last_err = None;

    for addr in addrs {
        match TcpStream::connect_addr(addr).await {
            Ok(stream) => return Ok(stream),
            Err(e) => last_err = Some(e),
        }
    }

   // ...
```

the cool thing is that it automatically resolves dns names which is something we need to do manually right now

File: msg-socket/src/sub/driver.rs [hunk 2]
```diff
@@ -157,12 +159,18 @@ where
...
-            Command::Connect { endpoint } => {
-                let id = self.options.auth_token.clone();
-                let transport = Arc::clone(&self.transport);
+            Command::Connect { mut endpoint } => {
+                // Some transport implementations (e.g. Quinn) can't dial an unspecified IP address, so replace
+                // it with localhost.
+                if endpoint.ip().is_unspecified() {
+                    // TODO: support IPv6
+                    endpoint.set_ip(IpAddr::V4(Ipv4Addr::LOCALHOST));
+                }
```
Comment (merklefruit) [inline msg-socket/src/sub/driver.rs:168]: will this work in a real network environment?

File: msg-transport/src/durable/session.rs [hunk 3]
```diff
@@ -225,6 +264,7 @@ pub struct DurableSession<Io: UnderlyingIo> {
     /// Optional layer stack. If this is `None`, newly connected (or reconnected) sessions will
     /// be passed through without processing.
     layer_stack: Option<Box<dyn Layer<Io> + Send>>,
+    should_reconnect: bool,
```
Comment (merklefruit) [inline msg-transport/src/durable/session.rs:267]:
```suggestion
    layer_stack: Option<Box<dyn Layer<Io> + Send>>,
    /// If the session should be reconnected upon an unexpected disconnection
    should_reconnect: bool,
```

**Review by merklefruit — APPROVED** [review]
Wonderful

File: msg-socket/src/pub/socket.rs (hunk 1 above)
Comment (mempirate) [reply msg-socket/src/pub/socket.rs:69]: Was also thinking about this, will make issue w.r.t. QoL improvements

**Comment by mempirate** [comment]
### Merge activity

* **Jan 18, 9:01 AM**: @mempirate started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/chainbound/msg-rs/48).

**Pull request merged**